File: src/py3/shell_sort.py
from math import floor
from typing import List, Any


def get_gaps(size: int) -> List[int]:
    temp = size // 2
    gaps = [temp]
    while temp > 1:
        temp = temp // 2
        gaps.append(temp)

    return gaps


def get_gaps_knuth(size: int) -> List[int]:
    temp = 1
    gaps = [temp]
    while temp < size:
        temp = temp * 3 + 1
        gaps.append(temp)

    return gaps[::-1]


def get_gaps_ciura(size: int) -> List[int]:
    gaps_init = [1, 4, 10, 23, 57, 132, 301, 701, 1750]
    temp = 1
    gaps = []
    i = 0
    while temp < size:
        temp = floor(temp * 2.25) if temp >= 1750 else gaps_init[i]
        gaps.append(temp)
        i += 1

    return gaps[::-1]


def shell_sort(seq: List[Any]) -> None:
    for gap in get_gaps_ciura(len(seq)):
        for i in range(gap, len(seq)):
            i_insert = i
            key = seq[i]
            while (i_insert >= gap) and (key < seq[i_insert - gap]):
                seq[i_insert] = seq[i_insert - gap]
                i_insert -= gap

            seq[i_insert] = key


if __name__ == "__main__":
    sorting_fn = shell_sort

    def is_sorted(seq: List[Any]) -> bool:
        for i in range(len(seq)-1):
            if seq[i] > seq[i+1]:
                return False
        return True

    # empty
    seq = []
    sorting_fn(seq)
    print(f"test_empty: {'passed' if is_sorted(seq) else 'failed'}.")

    # +
    seq = [16, 7, 9, 5, 65, 49, 37, 3, 28, 2, 21, 12, 4]
    sorting_fn(seq)
    print(f"test_int_1: {'passed' if is_sorted(seq) else 'failed'}.")

    # +, -
    seq = [-16, 7, -9, 5, -65, 49, -37, 3, -28, 2, -21, 12, -4]
    sorting_fn(seq)
    print(f"test_int_2: {'passed' if is_sorted(seq) else 'failed'}.")

    # +, -, =
    seq = [1, 2, -4, -7, 1, 9, -7, 2, 2, 6, 1, 12, 4]
    sorting_fn(seq)
    print(f"test_int_3: {'passed' if is_sorted(seq) else 'failed'}.")

    # string
    seq = ["p", "g", "i", "j", "65", "W", "K", "c", "B", "b", "21", "l", "d"]
    sorting_fn(seq)
    print(f"test_str: {'passed' if is_sorted(seq) else 'failed'}.")

File: src/rust/src/sort/shell_sort.rs
fn _get_gaps(size: usize) -> Vec<usize> {
    let mut temp = size / 2;
    let mut gaps = vec![temp];
    while temp > 1 {
        temp = temp / 2;
        gaps.push(temp);
    }
    gaps
}


fn _get_gaps_knuth(size: usize) -> Vec<usize> {
    let mut temp = 1;
    let mut gaps = vec![temp];
    while temp < size {
        temp = temp * 3 + 1;
        gaps.push(temp);
    }
    gaps.reverse();
    gaps
}


fn _get_gaps_ciura(size: usize) -> Vec<usize> {
    let gaps_init = vec![1, 4, 10, 23, 57, 132, 301, 701, 1750];
    let mut temp = 1;
    let mut gaps = Vec::new();
    let mut i = 0;
    while temp < size {
        if temp >= 1750 {
            temp = ((temp as f64 * 2.22).floor()) as usize;
        } else {
            temp = gaps_init[i];
        }
        gaps.push(temp);
        i += 1;
    }
    gaps.reverse();
    gaps
}


pub fn shell_sort<T>(seq: &mut [T])
where
    T: PartialOrd + Copy
{
    let gaps = _get_gaps_ciura(seq.len());

    for i in 0..gaps.len() {
        let gap = gaps[i];
        for j in gap..seq.len() {
            let mut i_insert = j;
            let key = seq[j];
            while (i_insert >= gap) && (key < seq[i_insert - gap]) {
                seq[i_insert] = seq[i_insert - gap];
                i_insert = i_insert - gap;
            }
            seq[i_insert] = key;
        }
    }
}
